### ISSUE TITLE
Support multiple entry points

### ISSUE THREAD

**Issue by tungld** [onnx/onnx-mlir#1183]
Currently, onnx-mlir has not yet fully supported multiple entry points in a model, e.g. We assume there is only a single entry point, that is `run_main_graph` and provide `omInputSignature()` and `omOutputSignature()` functions for querying the input/output of the entry point. However, we would like to extend these signature functions in order to support multiple entry points.

Example of a multi-entry-point code:
```mlir
module {
  func private @test_add(%arg0 : tensor<10xf32>, %arg1 : tensor<10xf32>) -> tensor<*xf32> {
    %0 ="onnx.Add"(%arg0, %arg1) : (tensor<10xf32>, tensor<10xf32>) -> tensor<*xf32>
    "std.return"(%0) : (tensor<*xf32>) -> ()
  }
  func private @test_relu(%arg0 : tensor<10xf32>) -> tensor<*xf32> {
    %0 ="onnx.Relu"(%arg0) : (tensor<10xf32>) -> tensor<*xf32>
    "std.return"(%0) : (tensor<*xf32>) -> ()
  }
  "onnx.EntryPoint"() {func = @test_add, numInputs = 2 : i32, numOutputs = 1 : i32, signature = ""} : () -> ()
  "onnx.EntryPoint"() {func = @test_relu, numInputs = 1 : i32, numOutputs = 1 : i32, signature = ""} : () -> ()
}
```

There are two approaches to supporting multiple entry points:
1. Reuse function names `omInputSignature` and `omOutputSignature`, and allow users to pass an entry point name to query its signature, e.g. `omInputSignature("run_test_add")`.
2. For each entry point name, e,g. `run_test_add`, create two functions `run_test_add_in_signature()` and `run_test_add_out_signature()` to query its signatures.

Though approach 1 reuses function names, we cannot reuse the function content and we need to generate complicated code, e.g. generating blocks and condition branches at LLVM level for searching the signature of the given entry point name.

Meanwhile, approach 2 does not reuse the function names but it is straightforward to implement. We simply use a dynamic name, e.g `run_test_add_in_signature` depending on the entry point name, instead of a fixed name `omInputSignature`. There is no change to the function content and no searching needed.

I prefer approach 2. However, I am a bit concerned about changing the function names since onnx-mlir users are using those names.

**Comment by gongsu832** [comment]
I don't know how much complexity option #1 adds to the internal code but from the backward compatibility point of view option #1 is preferred. Typically, you would designate a "default" entry point so that the existing `omInputSignature()` and `omOutputSignature()` would return signatures for the default entry point and add the new functions `omInputSignature(entry_point_name)` and `omOutputSignature(entry_point_name)`. This works well for C++ and Java. For C existing code will have to be changed since it does not support same function with different signature. But that's typically a small price to pay.

Also from a general code practice, hard code entry point name into the function name like option #2 is also less desirable since it would force users to write code like
```
if (entry_point_name == "test_add")
    run_test_add_in_signature();
else if (entry_point_name == "test_relu")
    run_test_relu_in_signature();
...
```
In other words, it forces users to write code that maps an entry point name to its corresponding signature function. This complexity should be contained in the runtime internal code. Since it's machine generated so less error prone. And you probably already have some logic to map an entry point name to its corresponding signature functions anyway in order to implement option #2. Since the runtime is internally implemented in C++, it shouldn't be too hard to do the mapping with the 
`std::map` construct.

**Comment by tungld** [comment]
@gongsu832 so, looks like approach 2 eases the burden on users. `omInputSignature` is a function emitted in LLVMIR, so it cannot be overloaded. If we provide only `omInputSignature(entry_point_name)`, is it fine for C, C++, and Java interfaces? They should know the default entry point is "run_main_graph" and pass it to `omInputSignature`.

**Comment by gongsu832** [comment]
@tungld You mean approach 1 eases the burden on users? It should be fine if we provide only `omInputSignature(entry_point_name)`. C/C++ callers will have to make a small change to existing code. Java has its own wrapper `inputSignature()` around `omInputSignature()` so we can just add the new `inputSignature(entry_point_name)` wrapper without requiring Java code change.

**Comment by tungld** [comment]
> @tungld You mean approach 1 eases the burden on users?

Yes, my mistake. I mean approach 1.

**Comment by AlexandreEichenberger** [comment]
@tungld,

I think we must be very clear on what the "entry point" are. It may be self evident to us now, but this is important for the user as they are exposed to these names. Is it correct that these "entry point" are actually the symbol of the compute function? If so, they are exposed to the user as if you have a statically linked compute function, like in many of our examples, we expect the user to define and then call `entry_point(inputTensorList)`, and if the user is working with dynamic libraries, the user is expected to do a `dlsym(entry_point).

I agree with @gongsu832 that we should have functions to get the signatures that passes the entry point name. But this starts to feel like we have a data structure for each entry point.

```C
struct EntryPointStruct {
  char *entryPointName;
  char *inputSignature;
  char *outputSignature;
}
```

That feel preferable to having hardwired functions to get the signatures, as if we have multiple entry point, we must start to enumerate/generate to get the right function for a solution 1

```C
if (entry_point_name == "test_add")
    run_test_add_in_signature();
else if (entry_point_name == "test_relu")
    run_test_relu_in_signature();
```

(since the name are constants, probably test the address of the constant is enough).
